Make it easier to tween properties with custom easings

### Describe the project you are working on

Godot

### Describe the problem or limitation you are having in your project

There were a couple proposals about tweening using curves and the topic seems to re-appear from time to time. The reason it wasn't implemented yet is that it's easy to achieve it using MethodTweener:
```GDScript
tween.tween_method(func(v): position.y = curve.sample_baked(v), 0, curve.get_baked_length(), 1.0)
```
There is however a rather prominent problem with that: the property you want to animate has to be hard-coded in the method (in this case the method is a lambda). Alternative is passing object and property to the method, but then we just end up with wannabe PropertyTweener that doesn't provide the convenient methods like `as_relative()` or `from()`.

Another thing is that it's verbose. If you want to interpolate curve, you don't care about it's length. In the above example the length can be moved inside method and the method could run from 0 to 1 instead, but you'd still need to provide the 0-1 range to the method, which is a chore.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a new method to PropertyTweener. It could be called `set_custom_method()` and would take a Callable. Then instead of doing easing, the tweener would call the method and use its output for the value. For simplicity, the method would be normalized to take 0-1 range as argument and return 0-1. Then the code above can be changed to:
```GDScript
tween.tween_property(self, "position:y", 300.0, 1.0).set_custom_method(interpolate_curve.bind(curve))
```
The `interpolate_curve` would be:
```GDScript
func interpolate_curve(v, curve):
	return curve.sample_baked(remap(v, 0, curve.get_baked_length()))
```
What's better, you can wrap this method to make it more convenient to use:
```GDScript
TweenUtils.tween_curve(tween, self, "position:y", 300, 1.0, curve)
```
Which makes it easy to distribute as an addon.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Add a Callable `custom_method` to PropertyTweener, which would be called inside `step()` instead of running equation. The equation could be run on the method's parameter (otherwise `set_trans()` and `set_custom_method()` would cancel out each other).

### If this enhancement will not be used often, can it be worked around with a few lines of script?

It can be worked around, but the workaround is less convenient. This is about convenience.
Also tweening with custom curves is a frequently requested feature. The feature suggested here would make it easy to implement curves as an addon.

### Is there a reason why this should be core and not an add-on in the asset library?

This makes it easier to create addons.

While the PR isn't merged, I've made a small script that makes `MethodTweener` simulate `PropertyTweener` with a custom interpolation.
It's still very limited by a lot of factors, but it should be flexible enough for the most basic use cases while reducing verbosity.

> tween_utils.gd
```gdscript
# Add this script to AutoLoad as TweenUtils
# Can only set full properties:
#  [OK]    "position"
#  [ERROR] "position:x"

extends Node

# Tween a property using a custom interpolation method
func interpolate_property(tween : Tween, interpolator_method : Callable, object : Object, property : StringName, final_val : Variant, duration : float) -> MethodTweener:
	var inner_method := inner_interpolator.bind(interpolator_method, object, property, object.get(property), final_val)
	return tween.tween_method(inner_method, 0.0, 1.0, duration)

# Shortcut to directly use a curve instead of a method
func interpolate_curve_property(tween : Tween, interpolation_curve : Curve, object : Object, property : StringName, final_val : Variant, duration : float) -> MethodTweener:
	var interpolator_method := func(value, curve) -> float: return curve.sample_baked(value)
	interpolator_method = interpolator_method.bind(interpolation_curve)
	return interpolate_property(tween, interpolator_method, object, property, final_val, duration)

# Inner method, used by Tween.tween_method()
func inner_interpolator(tween_value : float, interpolator_method : Callable, object : Object, property: StringName, from : Variant, to : Variant):
	var interpolated_value = interpolator_method.call(tween_value)
	var interpolated_property
	
	if(interpolated_value is float):
		interpolated_property = (to - from) * interpolated_value + from
	else:
		interpolated_property = (to - from) * tween_value + from
		push_error(str(interpolator_method.get_object()) + "." + interpolator_method.get_method() + "() method does not return a float value")
	
	object.set(property, interpolated_property)

# [Example]
# TweenUtils.interpolate_property(tween, my_interpolator, self, "position", target.position, duration)
#
# func my_interpolator(tween_value : float) -> float:
#	return my_curve.sample_baked(tween_value)
```

--- a/doc/classes/PropertyTweener.xml
+++ b/doc/classes/PropertyTweener.xml
@@ -43,6 +43,25 @@
 				[/codeblock]
 			</description>
 		</method>
+		<method name="set_custom_interpolator">
+			<return type="PropertyTweener" />
+			<param index="0" name="interpolator_method" type="Callable" />
+			<description>
+				Allows interpolating the value with a custom easing function. The provided [param interpolator_method] will be called with a value ranging from [code]0.0[/code] to [code]1.0[/code] and is expected to return a value within the same range (values outside the range can be used for overshoot). The return value of the method is then used for interpolation between initial and final value. Note that the parameter passed to the method is still subject to the tweener's own easing.
+				[b]Example:[/b]
+				[codeblock]
+				@export var curve: Curve
+
+				func _ready():
+				    var tween = create_tween()
+				    # Interpolate the value using a custom curve.
+				    tween.tween_property(self, "position:x", 300, 1).as_relative().set_custom_interpolator(tween_curve)
+
+				func tween_curve(v):
+				    return curve.sample_baked(v)
+				[/codeblock]
+			</description>
+		</method>
 		<method name="set_delay">
 			<return type="PropertyTweener" />
 			<param index="0" name="delay" type="float" />

--- a/scene/animation/tween.cpp
+++ b/scene/animation/tween.cpp
@@ -531,6 +531,11 @@ Ref<PropertyTweener> PropertyTweener::set_ease(Tween::EaseType p_ease) {
 	return this;
 }
 
+Ref<PropertyTweener> PropertyTweener::set_custom_interpolator(const Callable &p_method) {
+	custom_method = p_method;
+	return this;
+}
+
 Ref<PropertyTweener> PropertyTweener::set_delay(double p_delay) {
 	delay = p_delay;
 	return this;
@@ -581,7 +586,23 @@ bool PropertyTweener::step(double &r_delta) {
 
 	double time = MIN(elapsed_time - delay, duration);
 	if (time < duration) {
-		target_instance->set_indexed(property, tween->interpolate_variant(initial_val, delta_val, time, duration, trans_type, ease_type));
+		if (custom_method.is_valid()) {
+			const Variant t = tween->interpolate_variant(0.0, 1.0, time, duration, trans_type, ease_type);
+			const Variant *argptr = &t;
+
+			Variant result;
+			Callable::CallError ce;
+			custom_method.callp(&argptr, 1, result, ce);
+			if (ce.error != Callable::CallError::CALL_OK) {
+				ERR_FAIL_V_MSG(false, "Error calling custom method from PropertyTweener: " + Variant::get_callable_error_text(custom_method, &argptr, 1, ce) + ".");
+			} else if (result.get_type() != Variant::FLOAT) {
+				ERR_FAIL_V_MSG(false, vformat("Wrong return type in PropertyTweener custom method. Expected float, got %s.", Variant::get_type_name(result.get_type())));
+			}
+
+			target_instance->set_indexed(property, Animation::interpolate_variant(initial_val, final_val, result));
+		} else {
+			target_instance->set_indexed(property, tween->interpolate_variant(initial_val, delta_val, time, duration, trans_type, ease_type));
+		}
 		r_delta = 0;
 		return true;
 	} else {
@@ -609,6 +630,7 @@ void PropertyTweener::_bind_methods() {
 	ClassDB::bind_method(D_METHOD("as_relative"), &PropertyTweener::as_relative);
 	ClassDB::bind_method(D_METHOD("set_trans", "trans"), &PropertyTweener::set_trans);
 	ClassDB::bind_method(D_METHOD("set_ease", "ease"), &PropertyTweener::set_ease);
+	ClassDB::bind_method(D_METHOD("set_custom_interpolator", "interpolator_method"), &PropertyTweener::set_custom_interpolator);
 	ClassDB::bind_method(D_METHOD("set_delay", "delay"), &PropertyTweener::set_delay);
 }
 
@@ -685,7 +707,7 @@ bool CallbackTweener::step(double &r_delta) {
 		Callable::CallError ce;
 		callback.callp(nullptr, 0, result, ce);
 		if (ce.error != Callable::CallError::CALL_OK) {
-			ERR_FAIL_V_MSG(false, "Error calling method from CallbackTweener: " + Variant::get_callable_error_text(callback, nullptr, 0, ce));
+			ERR_FAIL_V_MSG(false, "Error calling method from CallbackTweener: " + Variant::get_callable_error_text(callback, nullptr, 0, ce) + ".");
 		}
 
 		finished = true;
@@ -765,7 +787,7 @@ bool MethodTweener::step(double &r_delta) {
 	Callable::CallError ce;
 	callback.callp(argptr, 1, result, ce);
 	if (ce.error != Callable::CallError::CALL_OK) {
-		ERR_FAIL_V_MSG(false, "Error calling method from MethodTweener: " + Variant::get_callable_error_text(callback, argptr, 1, ce));
+		ERR_FAIL_V_MSG(false, "Error calling method from MethodTweener: " + Variant::get_callable_error_text(callback, argptr, 1, ce) + ".");
 	}
 
 	if (time < duration) {

--- a/scene/animation/tween.h
+++ b/scene/animation/tween.h
@@ -197,6 +197,7 @@ public:
 	Ref<PropertyTweener> as_relative();
 	Ref<PropertyTweener> set_trans(Tween::TransitionType p_trans);
 	Ref<PropertyTweener> set_ease(Tween::EaseType p_ease);
+	Ref<PropertyTweener> set_custom_interpolator(const Callable &p_method);
 	Ref<PropertyTweener> set_delay(double p_delay);
 
 	void set_tween(const Ref<Tween> &p_tween) override;
@@ -222,6 +223,7 @@ private:
 	double duration = 0;
 	Tween::TransitionType trans_type = Tween::TRANS_MAX; // This is set inside set_tween();
 	Tween::EaseType ease_type = Tween::EASE_MAX;
+	Callable custom_method;
 
 	double delay = 0;
 	bool do_continue = true;

